File: 17/B17.java
public class B17 {
	public static void print(int[] ll){
		print(ll, 0);
	}

	public static void print(int[] ll, int index){
		System.out.print(index + " ");
		index = ll[index];
		if (index == 0){
			System.out.println();
			return;
		}
		print(ll, index);
	}
	
	public static void main(String[] args){
		int step = 354;
		int stop = 50000000;
		
		int[] ll = new int[stop + 1];
		int ptr = 0;

		for (int i = 1; i <= stop; i++){

			if (i % 100000 == 0){
				System.out.println(i);
			}

			for (int j = 0; j < step; j++){
				ptr = ll[ptr];
			}
			ll[i] = ll[ptr];
			ll[ptr] = i;
			ptr = i;
		}
		
		System.out.print(ll[0]);
	}
}

File: 17/b17.py
#!/usr/bin/env python

step = 354
stop = 50000000

arr = {}
arr[0] = 0
ptr = 0

#def next(n):
#    if n == 0:
#        return ptr
#    else:
#        ptr = arr[ptr]
#        return next(n-1)

def pd(d, index):
    print(index, end=" ")
    if d[index] == 0:
        print()
        return
    else:
        pd(d, d[index])

for i in range(1,stop+1):
    if i%100000 == 0:
        print(i)
    for j in range(0,step):
        ptr = arr[ptr]
    arr[ptr], arr[i] = i, arr[ptr]
    ptr = arr[ptr]

print(arr[0])

